src/core_plugins/choropleth/public/choropleth_controller.js

```diff
@@ -0,0 +1,157 @@
+      choroplethLayer.setMetrics(results, metricsAgg);
+      setTooltipFormatter();
+
+      kibanaMap.useUiStateFromVisualization($scope.vis);
+      kibanaMap.resize();
+      $element.trigger('renderComplete');
+    });
+  });
+
+  $scope.$watch('vis.params', (visParams) => {
+    kibanaMapReady.then(() => {
+      if (!visParams.selectedJoinField) {
+        visParams.selectedJoinField = visParams.selectedLayer.fields[0];
+      }
+
+      updateChoroplethLayer(visParams.selectedLayer.url);
+      choroplethLayer.setJoinField(visParams.selectedJoinField.name);
+      choroplethLayer.setColorRamp(truncatedColorMaps[visParams.colorSchema]);
+      setTooltipFormatter();
+
+      kibanaMap.setShowTooltip(visParams.addTooltip);
+      kibanaMap.setLegendPosition(visParams.legendPosition);
+      kibanaMap.useUiStateFromVisualization($scope.vis);
+      kibanaMap.resize();
+      $element.trigger('renderComplete');
+    });
+  });
+
+
+  function setTooltipFormatter() {
+    const metricsAgg = _.first($scope.vis.aggs.bySchemaName.metric);
+    if ($scope.vis.aggs.bySchemaName.segment && $scope.vis.aggs.bySchemaName.segment[0]) {
+      const fieldName = $scope.vis.aggs.bySchemaName.segment[0].makeLabel();
+      choroplethLayer.setTooltipFormatter(tooltipFormatter, metricsAgg, fieldName);
+    } else {
+      choroplethLayer.setTooltipFormatter(tooltipFormatter, metricsAgg, null);
+    }
+  }
+
+  function updateChoroplethLayer(url) {
+
+    if (choroplethLayer && choroplethLayer.equalsGeoJsonUrl(url)) {
+      return;
+    }
+    kibanaMap.removeLayer(choroplethLayer);
+
+    const previousMetrics = choroplethLayer ? choroplethLayer.getMetrics() : null;
+    const previousMetricsAgg = choroplethLayer ? choroplethLayer.getMetricsAgg() : null;
+    choroplethLayer = new ChoroplethLayer(url);
+    if (previousMetrics && previousMetricsAgg) {
+      choroplethLayer.setMetrics(previousMetrics, previousMetricsAgg);
+    }
+    choroplethLayer.on('select', function (event) {
+      const aggs = $scope.vis.aggs.getResponseAggs();
+      const aggConfigResult = new AggConfigResult(aggs[0], false, event, event);
+      $scope.vis.listeners.click({ point: { aggConfigResult: aggConfigResult } });
+    });
+    choroplethLayer.on('styleChanged', function (event) {
+      if (event.mismatches.length > 0 && config.get('visualization:vectormap:showWarnings')) {
+        notify.warning(
+          `Could not show ${event.mismatches.length} ${event.mismatches.length > 1 ? 'results' : 'result'} on the map.`
+            + ` To avoid this, ensure that each term can be joined to a corresponding shape on that shape's join field.`
+            + ` Could not join terms: ${event.mismatches.join(',')}`
+        );
+      }
+    });
+    kibanaMap.addLayer(choroplethLayer);
+
+  }
+
+
+  function getValue(metricsAgg, bucket) {
+    let size = metricsAgg.getValue(bucket);
+    if (typeof size !== 'number' || isNaN(size)) {
+      try {
```

Comment:
I've been unable to find this code-path being hit, or other visualizations (besides tagcloud) doing something similar.